libp2p/stream_muxer/yamux/yamux.py

```diff
@@ -0,0 +1,270 @@
+from libp2p.peer.id import (
+    ID,
+)
+from libp2p.stream_muxer.exceptions import (
+    MuxedStreamError,
+)
+
+PROTOCOL_ID = "/yamux/1.0.0"
+TYPE_DATA = 0x0
+TYPE_WINDOW_UPDATE = 0x1
+TYPE_PING = 0x2
+TYPE_GO_AWAY = 0x3
+FLAG_SYN = 0x1
+FLAG_ACK = 0x2
+FLAG_FIN = 0x4
+FLAG_RST = 0x8
+HEADER_SIZE = 12
+
+
+class YamuxStream(IMuxedStream):
+    def __init__(self, stream_id: int, conn: "Yamux", is_initiator: bool) -> None:
+        self.stream_id = stream_id
+        self.conn = conn
+        self.is_initiator = is_initiator
+        self.closed = False
+
+    async def write(self, data: bytes) -> None:
+        if self.closed:
+            raise MuxedStreamError("Stream is closed")
+        header = struct.pack("!BBHII", 0, TYPE_DATA, 0, self.stream_id, len(data))
+        await self.conn.secured_conn.write(header + data)
+
+    async def read(self, n: int = -1) -> bytes:
+        return await self.conn.read_stream(self.stream_id, n)
+
+    async def close(self) -> None:
+        if not self.closed:
+            logging.debug(f"Closing stream {self.stream_id}")
+            header = struct.pack("!BBHII", 0, TYPE_DATA, FLAG_FIN, self.stream_id, 0)
+            await self.conn.secured_conn.write(header)
+            self.closed = True
+
+    async def reset(self) -> None:
+        if not self.closed:
+            logging.debug(f"Resetting stream {self.stream_id}")
+            header = struct.pack("!BBHII", 0, TYPE_DATA, FLAG_RST, self.stream_id, 0)
+            await self.conn.secured_conn.write(header)
+            self.closed = True
+
+    def set_deadline(self, ttl: int) -> bool:
+        return False
+
+    def get_remote_address(self) -> Optional[tuple[str, int]]:
+        """
+        Returns the remote address of the underlying connection.
+        """
+        # Delegate to the secured_conn's get_remote_address method
+        if hasattr(self.conn.secured_conn, "get_remote_address"):
+            remote_addr = self.conn.secured_conn.get_remote_address()
+            # Ensure the return value matches tuple[str, int] | None
+            if (
+                remote_addr is None
+                or isinstance(remote_addr, tuple)
+                and len(remote_addr) == 2
+            ):
+                return remote_addr
+            else:
+                raise ValueError(
+                    "Underlying connection returned an unexpected address format"
+                )
+        else:
+            # Return None if the underlying connection doesn’t provide this info
+            return None
+
+
+class Yamux(IMuxedConn):
+    def __init__(
+        self,
+        secured_conn: ISecureConn,
+        peer_id: ID,
+        is_initiator: Optional[bool] = None,
+    ) -> None:
+        self.secured_conn = secured_conn
+        self.peer_id = peer_id
+        self.is_initiator_value = (
+            is_initiator if is_initiator is not None else secured_conn.is_initiator
+        )
+        self.next_stream_id = 1 if self.is_initiator else 2
```

Comment:
I was curious about this. A comment referring to how this is from yamux spec def for stream_id would be helpful.

Reply:
great!